### PR TITLE
Fix BOX downscale dropping boundary source pixels

### LINKED ISSUE #9939: BOX filter missed some edge-case pixels
### What did you do?

I down-sample an image with BOX filter.

Here is the image:

[table-51.png](https://github.com/user-attachments/assets/fd51bff0-caf2-4c39-b002-7872ccb8de82) (width="2755" height="1837")

---

### What did you expect to happen?

All pixels are counted while down-sampling.

The result should look like this (688x459, 459x306):

Image

[box-js-v3-table-51.png-688x459-49c214b-pil-like.png](https://github.com/user-attachments/assets/4a8c14c4-e7ab-4d31-80e2-0d2f6df4e994)

\---

Image

[box-js-v3-table-51.png-459x306-20d1f1f-pil-like.png](https://github.com/user-attachments/assets/4edbcc87-b42a-44a0-8349-ac2350edba80)

---

### What actually happened?

Some edge-case pixels were skipped.

\---

Downscaled `4` times to 688x459:

Image

[box-pil-table-51.png-688x459-49c7ade.png](https://github.com/user-attachments/assets/38dd97ca-013c-4782-8e82-d3f9b75d14f1)

missed a central vertical line.

\---

Downscaled `6` times to 459x306:

Image

[box-pil-table-51.png-459x306-20d5e62.png](https://github.com/user-attachments/assets/04a98318-4bab-4090-bf72-ed598ec7e0e2)

missed a central horizontal line.

---

### What are your OS, Python and Pillow versions?

* OS: Win 10
* Python: 3.12
* Pillow: 12.3.0

```python
file_path = "./table-51.png"
scale = 4  # scale(-down)

image = Image.open(file_path)
image_gray = image.convert("L")
wh = (int(image.width / scale), int(image.height / scale))
image = image_gray.resize(wh, Image.Resampling.BOX)
```

---

_Search words: off-by-one error, range rounding._

### PR DESCRIPTION
**Pull request by Dmao233**
Fixes #9939.

Changes proposed in this pull request:

 * BOX downscale assigns a source pixel whose center lands on an output bin edge to exactly one destination pixel
 * Add regression tests for the 13→6 boundary case and the issue sizes (2755→688, 1837→306)
 * Note the fix in `docs/releasenotes/13.0.0.rst`

### CODE REVIEW

File: docs/releasenotes/13.0.0.rst [hunk 1]
```diff
@@ -108,6 +108,15 @@ Two new filters are available for :py:meth:`~PIL.Image.Image.resize` and
...
 =============
 
+BOX downscale includes boundary source pixels
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+:data:`PIL.Image.Resampling.BOX` downscaling could skip a source pixel whose
+center landed on an output bin edge, due to range rounding and the discontinuous
+box kernel. Those pixels are now assigned to exactly one output pixel, matching
+the documented "each source pixel contributes to one destination pixel" behavior.
+:issue:`9939`
```
Comment (radarhere) [inline docs/releasenotes/13.0.0.rst]: We don't usually link to issues like this. Is there something you think the previous text doesn't sufficiently explain?

File: Tests/test_image_resample.py [hunk 2]
```diff
@@ -274,6 +274,29 @@ def test_box_filter_correct_range(self) -> None:
...
+
+        im = Image.new("L", (1, 13), 255)
+        im.putpixel((0, 6), 0)
+        out = im.resize((1, 6), Image.Resampling.BOX)
+        assert out.get_flattened_data() == (255, 255, 170, 255, 255, 255)
+
+        im = Image.new("L", (2755, 1), 255)
+        im.putpixel((1377, 0), 0)
+        out = im.resize((688, 1), Image.Resampling.BOX)
+        assert min(out.get_flattened_data()) < 255
```
Comment (radarhere) [inline Tests/test_image_resample.py]:
```suggestion
        assert not all(value == 255 for value in out.get_flattened_data())
```
The Lint job is currently failing. See what you think of this as a solution.

File: Tests/test_image_resample.py [hunk 3]
```diff
@@ -274,6 +274,29 @@ def test_box_filter_correct_range(self) -> None:
...
+
+        im = Image.new("L", (1, 1837), 255)
+        im.putpixel((0, 918), 0)
+        out = im.resize((1, 306), Image.Resampling.BOX)
+        assert min(out.get_flattened_data()) < 255
```
Comment (radarhere) [inline Tests/test_image_resample.py]:
```suggestion
        assert not all(value == 255 for value in out.get_flattened_data())
```

File: docs/releasenotes/13.0.0.rst (hunk 1 above)
Comment (Dmao233) [reply docs/releasenotes/13.0.0.rst]: No. Dropped in b43c6700.

File: Tests/test_image_resample.py (hunk 2 above)
Comment (Dmao233) [reply Tests/test_image_resample.py]: Took this in b43c6700.

File: Tests/test_image_resample.py (hunk 3 above)
Comment (Dmao233) [reply Tests/test_image_resample.py]: Took this in b43c6700.

**Comment by akx** [comment]
cc @AlttiRi as the reporter of #9939, you said in that issue you have a BOX down-sampling implementation of your own, so I suspect you have more domain knowledge, can you take a look at this? :)

File: src/libImaging/Resample.c [hunk 4]
```diff
@@ -272,24 +279,42 @@ precompute_coeffs(
         return 0;
     }
 
+    /* BOX maps each source pixel to exactly one output (docs: identical
+       weights). When downscaling, (int)(center ± support + 0.5) and the
```
Comment (radarhere) [inline src/libImaging/Resample.c]: Is 'docs: identical weights' intended to cause people to search and find https://pillow.readthedocs.io/en/stable/handbook/concepts.html#filters?

Comment (Dmao233) [reply src/libImaging/Resample.c]: No. Meant equal weights for BOX. Reworded in 097e00f7d.

**Comment by radarhere** [comment]
I've created https://github.com/Dmao233/Pillow/pull/2 with a suggestion.

**Comment by Dmao233** [comment]
Took the Simplify code suggestion from https://github.com/Dmao233/Pillow/pull/2.

**Comment by Dmao233** [comment]
Merged main; releasenotes conflict in 13.0.0.rst resolved (kept BOX note). e7be905.
